### DISCUSSION TITLE
Rename `isPromise` to `isPromiseLike` or `isThenable`

### DISCUSSION THREAD

**Discussion by aleclarson** [radashi-org/radashi#73] [Ideas]
Currently, `isPromise` is checking for a `then` method, but it should really be doing `Object.prototype.toString.call(value) === "[object Promise]"` instead.

The question is, do we rename the current implementation to `isPromiseLike` or `isThenable`?

**Comment by Minhir** [comment]
What type should it return?

**Reply by aleclarson** [reply]
Probably `value is { then: Function }`

**Reply by Minhir** [reply]
In such case `isThenable` looks like the perfect name

**Reply by aleclarson** [reply]
TypeScript has a built-in `PromiseLike<T>` type we could use, *but* there's no real way to safely ensure a `then` method has the signature of `then<TResult1 = T, TResult2 = never>(onfulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null): PromiseLike<TResult1 | TResult2>`.

Also, I checked and `value is { then: Function }` properly narrows the type `string | PromiseLike<string> | Promise<string>` to be `PromiseLike<string> | Promise<string>`.

**Reply by anion155** [reply]
> Also, I checked and `value is { then: Function }` properly narrows the type `string | PromiseLike<string> | Promise<string>` to be `PromiseLike<string> | Promise<string>`.

It will work for union types with `Promise<unknown>`, but it would not work for something like `unknown`. I'd say it's pretty safe to assume that anything that has `then` function field is `PromiseLike<unknown>`

**Reply by aleclarson** [reply]
@anion155 There are workarounds for that, but I'm thinking using `isPromiseLike` as a name and `value is PromiseLike<unknown>` as the return type is the most pragmatic solution. Like you said, reserving `then` methods for promise-like functionality is a best practice at this point.
